Cabal/doc/nix-local-build.markdown

```diff
@@ -0,0 +1,364 @@
+% Cabal User Guide: Nix-style local builds
+
+`cabal new-build`, also known as Nix-style local builds, is a new
+command inspired by Nix that comes with cabal-install 1.24. Nix-style
+local builds combine the best of non-sandboxed and sandboxed Cabal:
+
+1. Like sandboxed Cabal today, we build sets of independent local
+   packages deterministically and independent of any global state.
+   new-build will never tell you that it can't build your package
+   because it would result in a "dangerous reinstall." Given a
+   particular state of the Hackage index, your build is completely
+   reproducible. For example, you no longer need to compile packages
+   with profiling ahead of time; just request profiling and new-build
+   will rebuild all its dependencies with profiling automatically.
+
+2. Like non-sandboxed Cabal today, builds of external packages are
+   cached globally, so that a package can be built once, and then reused
+   anywhere else it is also used. No need to continually rebuild
+   dependencies whenever you make a new sandbox: dependencies which can
+   be shared, are shared.
+
+Nix-style local builds work with all versions of GHC supported by
+cabal-install 1.24, which currently is GHC 7.0 and later.
+
+Some features described in this manual are not implemented.  If you
+need them, please give us a shout and we'll prioritize accordingly.
+
+# Quickstart #
+
+Suppose that you are in a directory containing a single Cabal package
+which you wish to build.  You can configure and build it using Nix-style
+local builds with this command (configuring is not necessary):
+
+~~~~~~~~~~~~~~~~
+cabal new-build
+~~~~~~~~~~~~~~~~
+
+To open a GHCi shell with this package, use this command:
+
+~~~~~~~~~~~~~~~~
+cabal new-repl
+~~~~~~~~~~~~~~~~
+
+## Developing multiple packages ##
+
+Many Cabal projects involve multiple packages which need to be built
+together.  To build multiple Cabal packages, you need to first create a
+`cabal.project` file which declares where all the local package
+directories live. For example, in the Cabal repository, there is a root
+directory with a folder per package, e.g., the folders `Cabal` and
+`cabal-install`. The `cabal.project` file specifies each folder
+as part of the project:
+
+~~~~~~~~~~~~~~~~
+packages: Cabal/
+          cabal-install/
+~~~~~~~~~~~~~~~~
+
+The expectation is that a `cabal.project` is checked into your source
+control, to be used by all developers of a project.  If you need to
+make local changes, they can be placed in `cabal.project.local`
+(which should not be checked in.)
+
+Then, to build every component of every package, from the top-level
+directory, run the command:
+
+~~~~~~~~~~~~~~~~
+cabal new-build
```

Comment:
I think I'm going to argue for us to use 'all' here and to stick with the default being just the package in the current dir. Perhaps I'll implement it at MuniHac this weekend...